### ISSUE TITLE
Missing required failure message for CollectionInputFilter

### ISSUE THREAD

**Issue by Maks3w** [zendframework/zend-inputfilter#94]
CollectionInputFilter fails without error message when is required but data is empty or not set.

**Comment by n3vrax** [comment]
Is this fixed? Using the latest version, and it seems indeed that the collection input filter does not have some sort of default message when required is true and an empty array is given. Need to extend the input filter for now.

**Comment by developer-devPHP** [comment]
Issue is still exists. @weierophinney  can you confirm this?

**Comment by developer-devPHP** [comment]
Any updates on this? Problem still exists.

**Comment by weierophinney** [comment]
Please provide us with a minimal coffee sample that demonstrates the issue; right now, we'd be doing guess work.

**Comment by developer-devPHP** [comment]
# Example Input Filter
```php
$factory     = new \Zend\InputFilter\Factory();
$inputFilter = $factory->createInputFilter(
[
    'collection-element' => [
		'type'         => \Zend\InputFilter\CollectionInputFilter::class,
		'required'     => true,
		'input_filter' => [
			'type'  => [
				'required'     => true,
				'validators' => [
					[
						'name'    => \Zend\Validator\Between::class,
						'options' => [
							'min' => 50,
							'max' => 100,
						],
					],
				],
			],
			'price' => [
				'required'     => true,
				'validators' => [
					[
						'name'    => \Zend\Validator\Between::class,
						'options' => [
							'min' => 50,
							'max' => 100,
						],
					],
				],
			],
		],
	],
]
);
```
## First Try, when you have a problem with empty error message
```php
$inputFilter->setData([]);
var_dump($inputFilter->isValid());
var_dump($inputFilter->getMessages());
```
## Second Try with Valid error messages if provided some data
```php
$inputFilter->setData([
	'collection-element' => [
		0 => [

		],
	],
]);
var_dump($inputFilter->isValid());
var_dump($inputFilter->getMessages());
```

**Comment by froschdesign** [comment]
Output is:

### First try
```
boolean false

array (size=1)
  'collection-element' => 
    array (size=0)
      empty
```

### Second try
```
false

array (size=1)
  'collection-element' => 
    array (size=1)
      0 => 
        array (size=2)
          'type' => 
            array (size=1)
              'isEmpty' => string 'Value is required and can't be empty' (length=36)
          'price' => 
            array (size=1)
              'isEmpty' => string 'Value is required and can't be empty' (length=36)
```

**Comment by froschdesign** [comment]
No error message is generated.

Compare the `Input` class and the `CollectionInputFilter` class in this case:

https://github.com/zendframework/zend-inputfilter/blob/a2962cf00f86a58f2192ba63b572606dacbe1c20/src/Input.php#L410-L415

https://github.com/zendframework/zend-inputfilter/blob/a2962cf00f86a58f2192ba63b572606dacbe1c20/src/CollectionInputFilter.php#L171-L213

**Comment by developer-devPHP** [comment]
Do you have any updates related this issue?

**Comment by weierophinney** [comment]
@developer-devPHP No, no updates at this time.

We actually need to address this by getting some feedback first.

InputFilters _do not have a concept of required/optional by default_. We tacked it on to the `CollectionInputFilter` (and, for the 2.8.0 release, the `OptionalInputFilter`, which allows marking an input filter as optional, vs. required, meaning it can validate if the data set is empty). However, when we did, the only thing we had it affect is whether or not an empty set is valid. In the case that we have an empty set, _there is currently no mechanism to provide a message **why**_.

Normally, if data is present, `getMessages()` will return a nested set of messages, one for each data set:

```php
0 => [
    'foo' => [ /* array of messages */]
],
1 => [
    'bar' => [ /* array of mesages */ ]
],
/* etc. */
```

The problem area is that the each element of the above array is expected to be an array of key/messages pairs. How, exactly, do we represent a message indicating that a data set was empty in a way that is either consistent in structure, or easily differentiated from the existing structure?

Some thoughts off the top of my head:

- Use the same format, but use a "meta key", such as `__required__`:
  ```php
  [
      0 => [
          '__required__' => [ /* 'values are "required", but none present' messages */ ],
      ],
  ]
  ```
- Use a string value for the set:
  ```php
  [
      0 => 'Some error message here',
  ],
  ```

We need to choose one, but I need input from users as to which one to choose.

Additionally, we need:

- A default message to display when a collection is required.
- A mechanism for setting a custom "required" message.
- Updates to the factories to allow setting that message.

If anybody would like to tackle this, I'd appreciate the help.
